### ISSUE TITLE
Getting duplicate "IllegalStateException: Duplicate key" exception when using POJO rules

### ISSUE THREAD

**Issue by mill5james** [rulebook-rules/rulebook#126]
When using POJO rules in RuleBook, I get an "java.lang.IllegalStateException: Duplicate key <key>" exception if I have a rule that return `RuleState.NEXT` from the `@Then` method.

There is a [Gist ](https://gist.github.com/mill5james/367202b9d50d4353e4aae8778598c9f9) with the files also.

Invoking the rules is simple
```
public class App {
  public static void main(String[] args) {
    RuleBookRunner rules = new RuleBookRunner("rules");
    NameValueReferableMap<String> facts = new FactMap<>();
    facts.put("key", new Fact<>("value"));
    rules.run(facts);
  }
}
```
And there are two rules that I would expect to run sequentially
```
@Rule(order = 1)
public class ReturnsNext {
  @Given("key") 
  String value;

  @When
  public boolean when() { return true; }

  @Then
  public RuleState then() { return RuleState.NEXT; }
}
```
```
@Rule(order = 2)
public class SimpleRule {
  @Given("key")
  String value;

  @When
  public boolean when() { return true; }

  @Then
  public RuleState then() { return RuleState.BREAK; }
}
```

When you run the application, you will see an `IllegalStateException` exception raised
```
11:22:45.383 [main] ERROR com.deliveredtechnologies.rulebook.model.GoldenRule - Error occurred when trying to evaluate rule!
java.lang.IllegalStateException: Duplicate key value
	at java.util.stream.Collectors.lambda$throwingMerger$0(Collectors.java:133)
	at java.util.HashMap.merge(HashMap.java:1254)
	at java.util.stream.Collectors.lambda$toMap$58(Collectors.java:1320)
	at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1625)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at com.deliveredtechnologies.rulebook.model.GoldenRule.invoke(GoldenRule.java:139)
	at com.deliveredtechnologies.rulebook.model.runner.RuleAdapter.invoke(RuleAdapter.java:178)
	at com.deliveredtechnologies.rulebook.model.AuditableRule.invoke(AuditableRule.java:94)
	at com.deliveredtechnologies.rulebook.model.rulechain.cor.RuleHandler.handleRequest(RuleHandler.java:30)
	at com.deliveredtechnologies.rulebook.model.rulechain.cor.RuleHandler.lambda$handleRequest$1(RuleHandler.java:34)
	at java.util.Optional.ifPresent(Optional.java:159)
	at com.deliveredtechnologies.rulebook.model.rulechain.cor.RuleHandler.handleRequest(RuleHandler.java:32)
	at com.deliveredtechnologies.rulebook.model.rulechain.cor.CoRRuleBook.lambda$run$2(CoRRuleBook.java:46)
	at java.util.Optional.ifPresent(Optional.java:159)
	at com.deliveredtechnologies.rulebook.model.rulechain.cor.CoRRuleBook.run(CoRRuleBook.java:44)
	at com.deliveredtechnologies.rulebook.model.runner.RuleBookRunner.run(RuleBookRunner.java:93)
	at App.main(App.java:12)
```

**Comment by Clayton7510** [comment]
Thanks for providing your code. Unfortunately, I am as of yet, unable to reproduce your issue. Attached below is my complete code with gradle build script. You can run it using ```gradle run```.

My output was as follows and there were no errors:

```
We're about to run facts through a rule
Yay! facts just ran through the rule
```

It's funny, there are actually unit tests that address this scenario here https://github.com/rulebook-rules/rulebook/blob/develop/rulebook-core/src/test/java/com/deliveredtechnologies/rulebook/model/runner/RuleBookRunnerTest.java.

Still, I am a little interested in what could have caused such an error. So, I'll keep looking into it.

[simple-break-pojo.zip](https://github.com/rulebook-rules/rulebook/files/1933641/simple-break-pojo.zip)

**Comment by Clayton7510** [comment]
Fair enough. I'll check it out a bit more see if we can't get this thing resolved.

**Comment by Clayton7510** [comment]
@mill5james it turns out this was a bug, and an interesting one at that. There was a basic assumption that the name of a fact is also the key of the fact when added to a NameValueReferableMap. When that  assumption was broken, it turns out weird things happen. So, now the name of the fact as the key (and what matches the fact) is overridden by the NameValueReferableMap key value for each fact. The fix is in v0.10-SNAPSHOT.

I think it's probably also time to stop eating these kind of exceptions and make them a bit more targeted (i.e. fact is needed by a rule and the fact doesn't exist).

Thanks for raising this issue!

**Comment by mill5james** [comment]
@Clayton7510  Fantastic! Just tested it and the bug is fixed in the snapshot. Thanks for the fast turn around.
